### DISCUSSION TITLE
Need help with depRules

### DISCUSSION THREAD

**Discussion by pdsavard** [softarc-consulting/sheriff#126] [Q&A]
I am not sure to understand well how to applie rule.
I got 2 domains that nned to be access anywhere (for now) from any tags. Shared and Shell.
this is my config:

````
export const sheriffConfig: SheriffConfig = {
  excludeRoot: true,
  version: 1,
  tagging: {
    'src/app/@fuse/<fuse>': 'domain:fuse',
    'src/app/shell/<shell>': 'domain:shell',
    'src/app/domains/shared/<shared>': 'domain:shared',

    'src/app/domains/<domain>': {
      'feature-<feature>': ['domain:<domain>', 'type:feature'],
      'ui-<ui>': ['domain:<domain>', 'type:ui'],
      data: ['domain:<domain>', 'type:data'],
      '+store': ['domain:<domain>', 'type:data'],
      util: ['domain:<domain>', 'type:util'],
      routes: ['domain:<domain>', 'type:routes'],
    },
  },
  depRules: {
    root: ['domain:shell', 'domain:shared', anyTag],
    'domain:shell': [({ to }) => to.startsWith('domain'), 'type:feature', 'type:data', 'type:service'],
    'domain:*': ['domain:shared', 'domain:shell', anyTag],
    'type:feature': ['type:ui', 'type:data', 'type:util', 'type:animation', 'type:service'],
    'type:ui': ['type:data', 'type:util'],
    'type:data': ['type:util', 'type:data'],
    'type:util': ['type:data'],
  },
};
````

But I still get lint error like:
  5:1  error  module \src\app\domains\KelvinV2\+store cannot access \src\app\domains\shared\utils. Tag type:data has no clearance for tags domain:shared  @softarc/sheriff/dependency-rule
  9:1  error  module \src\app\domains\KelvinV2\+store cannot access \src\app\shell\notification. Tag type:data has no clearance for tags domain:shell     @softarc/sheriff/dependency-rule

**Comment by rainerhahnekamp** [comment]
Hi @pdsavard, yes, you are probably one of the developers who stumbled upon that fix in v0.17.0

Every from tag needs to have a clearance. In your case, the domain type is cleared but not the type one.

Can you update the following:
```typescript
    'src/app/shell/<shell>': 'domain:shell',
    'src/app/domains/shared/<shared>': 'domain:shared',
```
 to 

```typescript
    'src/app/shell/<shell>': ['domain:shell', 'type:shell'],
    'src/app/domains/shared/<shared>': ['domain:shared', 'type:shared'],
```

And also update your `depRules` accordingly. So if `type:data` needs to have access to`type:shell` and `type:shared`, it would be

```typescript
'type:data': ['type:util', 'type:data', 'type:shell', 'type:shared']
```

Let me know, if that fixes the issue for you.

**Comment by pdsavard** [comment]
You are probably one of the developers who stumbled upon that fix in v0.17.0! YES! 
I was put on another project, so I pause this angular one. But now I come back and see that you have a new version. I updated to the 0.17 one and try to take control of Sheriff config.

Maybe I got an incorrect structure too:
root
- domains
--- Shared
--- Domain1
--- Domain2
- Shell
- some individual file like app.component 

If I understand your answer, if I want to manually tag the Shell folder I need to set the domain AND the type to something? 
So I tag:
````
    'src/app/@fuse/<fuse>': ['domain:fuse', 'type:fuse'],
    'src/app/shell/<shell>': ['domain:shell', 'type:shell'],
    'src/app/domains/shared/<shared>': ['domain:shared', 'type:shared'],
````
And then use type dep rules instead of domain? I am not sure to understand that part. I was thinking user something like:
all domain can access domain Shell or Share with any tag like:
````'domain:*': ['domain:shared', 'domain:shell', anyTag],````
instead I remove my domain:* rule and add the shared and Shell type where I need to access some element in these domains like
```` 'type:data': ['type:util', 'type:data', 'type:shell', 'type:shared'] ````

Do I understand correctly? Sorry to ask such of basic question! ;-)

Question 2: I a add the auto taging bloc like:
````
    'src/app/domains/<domain>': {
      'feature-<feature>': ['domain:<domain>', 'type:feature'],
      'ui-<ui>': ['domain:<domain>', 'type:ui'],
      data: ['domain:<domain>', 'type:data'],
      '+store': ['domain:<domain>', 'type:data'],
      util: ['domain:<domain>', 'type:util'],
      routes: ['domain:<domain>', 'type:routes'],
    },
````
Does this will interfere with my manually Shared tag just above?

Question 3: I lost the red squiggly warning in my file. Now I need to use ng lint to see Sheriff errors. Do I miss something after updating to 0.17?

**Comment by pdsavard** [comment]
So I continue my learning of the v0.17 but is not simple! ;-)

My config is now:
 ````
export const sheriffConfig: SheriffConfig = {
  excludeRoot: true, // remove deepImport check from root (index.ts)
  tagging: {
    'src/app/shell/<shell>': 'shell',
    'src/app/domains/shared/<shared>': 'shared',

    // automatic tag of domains
    'src/app/domains/<domain>': {
      'feature-<feature>': ['domain:<domain>', 'type:feature'],
      'ui-<ui>': ['domain:<domain>', 'type:ui'],
      data: ['domain:<domain>', 'type:data'],
      '+store': ['domain:<domain>', 'type:store'],
      util: ['domain:<domain>', 'type:util'],
      routes: ['domain:<domain>', 'type:routes'],
    },
  },
  depRules: {
    root: ['noTag', 'shell', 'shared', 'domain:*'],
    noTag: ['noTag', 'shell', 'shared'],
    shell: anyTag,
    shared: anyTag,

    'type:feature': ['type:ui', 'type:data', 'type:util', 'type:animation', 'type:service', 'type:store'],
    'type:ui': ['type:data', 'type:util', 'type:store'],
    'type:store': ['type:util', 'type:data'],
    'type:data': ['type:util'],
    'type:util': ['type:data'],
  },
};
````
But I still have some lint error that I do not understand. Like a file in auth domain that need to access domain:auth but failed:

````
C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\KelvinV2\feature-kelvinv2\kelvin-v2.component.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:KelvinV2' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\+store\auth.state.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\+store\auth.store.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\data\modPermResponse.model.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\feature-signin\signing.component.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\ui-auth\maintenance\maintenance.component.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\ui-auth\presence-server\presence-server.component.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule

C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\auth\ui-auth\singin-form\login-form.component.ts
  1:1  error  User Error: SH-002 - No dependency rule for tag 'domain:auth' found in sheriff.config.ts  @softarc/sheriff/dependency-rule
````

The goal is that my Shell and Share are acessible from anywhere with any tag (for the moment).
Then the Domains can access according to the type rules dependencies
then root or noTag can access anything (for now)

any idea?

**Reply by rainerhahnekamp** [reply]
OK, so you still need the dependency rules for the domain group. You have to create rules for every tag.

Based on what you want to achieve, here's my recommendation:

```typescript
 depRules: {
   root: ['noTag', 'shell', 'shared', 'domain:*'],
   noTag: ['noTag', 'shell', 'shared'],
   shell: anyTag,
   shared: anyTag,

  'domain:*': [sameDomain, 'shell', 'shared'],

   'type:feature': ['type:ui', 'type:data', 'type:util', 'type:animation', 'type:service', 'type:store', 'shell', 'shared'],
   'type:ui': ['type:data', 'type:util', 'type:store', 'shell', 'shared'],
   'type:store': ['type:util', 'type:data', 'shell', 'shared'],
   'type:data': ['type:util', 'shell', 'shared'],
   'type:util': ['type:data', 'shell', 'shared'],
 },
```

At the moment, `shell` and `shared` can access anything. When you become more acquainted with the config, you should change that.

Let me know, if it works.

You find more detailed information on the dependency rules at https://softarc-consulting.github.io/sheriff/docs/dependency-rules

**Comment by rainerhahnekamp** [comment]
@pdsavard, let's make sure that your configs works and then we can discuss your other questions in detail.

**Comment by pdsavard** [comment]
sameDomain seems to not been detected in the import { anyTag, sameDomain, SheriffConfig } from '@softarc/sheriff-core';

adding this 'domain:*': [sameDomain, 'shell', 'shared'],
But now I got more error! ;-)

For exemple: 
C:\Data\_Dev\Angular\KelvinV3-2\src\app\domains\KelvinV2\+store\k2.state.ts
  1:1  error  module \src\app\domains\KelvinV2\+store cannot access \src\app\domains\KelvinV2\data. Tag domain:KelvinV2 has no clearance for tags domain:KelvinV2, type:data  @softarc/sheriff/dependency-rule

But k2.state.ts are in +store folder of the same "KelvinV2" domain. In the dependency rule, I got this rule: 'type:store': ['type:util', 'type:data'], so normally the store can access the data one?

**Reply by rainerhahnekamp** [reply]
Could you please provide the complete sheriff.config.ts and some error messages?

**Reply by pdsavard** [reply]
Ok now this config work.  I replace you sameDomain with sameTag

````
export const sheriffConfig: SheriffConfig = {
  excludeRoot: true, // remove deepImport check from root (index.ts)
  tagging: {
    'src/app/shell/<shell>': 'shell',
    'src/app/domains/shared/<shared>': 'shared',

    // automatic tag of domains
    'src/app/domains/<domain>': {
      'feature-<feature>': ['domain:<domain>', 'type:feature'],
      'ui-<ui>': ['domain:<domain>', 'type:ui'],
      data: ['domain:<domain>', 'type:data'],
      '+store': ['domain:<domain>', 'type:store'],
      util: ['domain:<domain>', 'type:util'],
      routes: ['domain:<domain>', 'type:routes'],
    },
  },
  depRules: {
    root: ['noTag', 'shell', 'shared', 'domain:*'],
    noTag: ['noTag', 'shell', 'shared'],
    shell: anyTag,
    shared: anyTag,

    'domain:*': [sameTag, 'shell', 'shared'],

    'type:feature': ['type:ui', 'type:data', 'type:util', 'type:animation', 'type:service', 'type:store', 'shared'],
    'type:ui': ['type:data', 'type:util', 'type:store', 'shared'],
    'type:store': ['type:util', 'type:data', 'shared', 'shell'],
    'type:data': ['type:util', 'shared'],
    'type:util': ['type:data', 'type:store', 'shared'],
  },
};
````

Now I will work on more restrain on shared and shell.

**Comment by rainerhahnekamp** [comment]
Sorry, I didn't see that I typed `sameDoman`. Of course it should be `sameTag`. Apologies!
